internal/gps/cmd_unix.go

```diff
@@ -25,11 +25,15 @@ type cmd struct {
 }
 
 func commandContext(ctx context.Context, name string, arg ...string) cmd {
-	// Grab the caller's context and pass a derived one to CommandContext.
-	c := cmd{ctx: ctx}
-	ctx, cancel := context.WithCancel(ctx)
-	c.Cmd = exec.CommandContext(ctx, name, arg...)
-	c.cancel = cancel
+	// Create a one-off cancellable context for use by the CommandContext, in
+	// the event that we have to force a Process.Kill().
+	ctx2, cancel := context.WithCancel(context.Background())
```

Comment:
same comment I left in #1269; we should just not use a context here and just `c.Process.Kill` directly instead.

Reply:
i think that's racy, unfortunately. IIRC, the code that your original PR got rid of in favor of `exec.CommandContext()` had all these nasty workarounds, including passing around integer address pointers for atomic ops, in order to be able to safely call `c.Cmd.Process.Kill()` directly without any possibility of that call being made on a nil pointer.

i'll go back and double-check, but i know that the thing i really liked about using `exec.CommandContext` in the first place was that it gave us a safe way of invoking `Kill()` by taking advantage of timing information that is otherwise only accessible from within the `os/exec` package.

Reply:
Please do double check - as far as I know, there's no possibility of `Cmd.Process` being nil after `Start` returns successfully.

Reply:
Ah, so, it was a bit more complicated than that - while i think you're correct that there's no possibility of `Command.Process` being nil, it was `Command.ProcessState` that gave problems:

https://github.com/golang/dep/blob/e74b613dd0afb09049628c38f1fd46c7fc4745f5/internal/gps/cmd_unix.go#L29-L53

the goal there was avoiding sending a signal to a process that had already terminated, and that's the thing that's difficult to ascertain from anywhere other than the thread that's `Wait()`ing. all i can remember about why i went to those lengths is because i wanted to avoid returning an error to the broader system when that error was simply a case of our mismanagement of the process (signal-after-termination), rather than an actual error _from_ the process. yes, we could sniff the error, but...well, it's an unexported, potentially os-specific error there. not ideal.

(and yes, that implies that this system should also probably be returning the cancelation error, rather than [one from the underlying process](https://github.com/golang/dep/blob/98f625e28cccd2c0b2549f1219fffb86d11d50bd/internal/gps/cmd_unix.go#L74), when cancelation is instructed)

so yeah, i'd prefer we stick with the reuse of context for the kill semantics here, as it both insulates us from having to think about any of the above, and may also provide a more useful point for understanding the causal ordering of events here (signal send start, signal send complete, process terminated, etc.)

Reply:
I don't understand what you're saying here. The code in `os/exec` is no more resilient to any of these problems; it just ignores the error returned from `Process.Kill`.

Reply:
¯\\\_(ツ)_/¯ - maybe it's all equivalent. i don't have more time i can afford to invest in this right now, and figuring out if there was something else that motivated the original code. so - if you and @jmank88 can agree that just calling `Process.Kill()` is equivalent, then i'll relent, and we can just use that.

if not, this route feels less dragon-y to me, and without adequate time to dispel the fog, i've gotta follow my gut.

Reply:
I think I'm with @tamird here.  We could do this in another PR though, right?  I can put something up after this is merged for us to stare at some more in isolation.

Reply:
yeah, good point, can do it in a follow-up.

Reply:
I thought that's what #1269 is.